### PR TITLE
Major reorganization breaking deploy into build + deploy steps

### PR DESCRIPTION
**Pull request by nexpoehler**
## Step 1
`ansible-playbook --extra-vars @sample-extra-vars/env-vars.yml -i inv.ini playbooks/environment.yml`

This creates the environment in which the client application will run.

## Step 2
`ansible-playbook --extra-vars @sample-extra-vars/build-vars.yml -i inv.ini playbooks/build.yml`

This creates a build of the client application and prepares it for deployment on the servers.

## Step 3
`ansible-playbook --extra-vars @sample-extra-vars/deploy-vars.yml -i inv.ini playbooks/deploy.yml`

This will deploy a previously created build into the "live" application location.

### CODE REVIEW

File: playbooks/_buildapp.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+---
+
+- name: Build Application
+  hosts: fpm[0]
```
Comment (clwells) [inline playbooks/_buildapp.yml]: We're assuming here that this is a PHP app, which for now is 100% fine.  Just need to keep in the back of our heads that this playbook will ultimately be lang agnostic.

File: playbooks/files/run-build-hook.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+BUILD=$1
+
+if [[ -f $BUILD/.nexcess-app.yml ]]; then
+    cd $BUILD
+    cat $BUILD/.nexcess-app.yml | shyaml get-value hooks.build | bash -x
```
Comment (clwells) [inline playbooks/files/run-build-hook.sh]: interesting and ILI.  This makes hooks 100% ansible's problem :).

File: playbooks/tasks/build/prep.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+   - name: Create build base
+     file: 
+       path="{{ nex_app_build_base }}"
+       state="directory"
+       mode="0711"
+       setype="_default"
+       owner="{{ nex_app_unix_user }}"
+       group="{{ nex_app_unix_group }}"
+   
+   - name: Create build code base
```
Comment (clwells) [inline playbooks/tasks/build/prep.yml]: title case

File: playbooks/tasks/build/prep.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+   
+   - name: Set build code base fact
+     set_fact:
+       nex_app_build_code_base="{{ nex_app_build_base }}/{{ nex_app_build_id }}/code"
+   
+   - name: Display Code Base
+     debug:
+       msg="nex_app_build_code_base = {{ nex_app_build_code_base }}"
+   
+   - name: Create build base
```
Comment (clwells) [inline playbooks/tasks/build/prep.yml]: title case

File: playbooks/tasks/build/prep.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+     shell: setfacl -m u:{{ nex_app_unix_user }}:x $(dirname $SSH_AUTH_SOCK)
+     become: true
+     become_flags: "-E"
+   
+   - name: Give user access to SSH_AUTH_SOCK
+     shell: setfacl -m u:{{ nex_app_unix_user }}:rwx $SSH_AUTH_SOCK
+     become: true
+     become_flags: "-E"
+   
+   - name: Setup secontext for build base
```
Comment (clwells) [inline playbooks/tasks/build/prep.yml]: title case

File: playbooks/tasks/build/prep.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
+---
+
+- block:
+   - name: Give user access to SSH_AUTH_SOCK dir
+     shell: setfacl -m u:{{ nex_app_unix_user }}:x $(dirname $SSH_AUTH_SOCK)
+     become: true
+     become_flags: "-E"
+   
+   - name: Give user access to SSH_AUTH_SOCK
```
Comment (clwells) [inline playbooks/tasks/build/prep.yml]: title case

File: playbooks/tasks/build/prep.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
+---
+
+- block:
+   - name: Give user access to SSH_AUTH_SOCK dir
```
Comment (clwells) [inline playbooks/tasks/build/prep.yml]: title case

File: playbooks/tasks/build/bundle.yml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+---
+- block:
+   - name: Bundle Build
+     archive:
+       path="{{ nex_app_build_base }}/{{ nex_app_build_id }}"
+       dest="{{ nfs_shared_path }}/{{ nex_app_build_id }}.tgz"
+       setype=yes
+     register: bundle_result
+ 
+   - name: Set fact for local bundle
```
Comment (clwells) [inline playbooks/tasks/build/bundle.yml]: title case

File: playbooks/deploy.yml [hunk 9]
```diff
@@ -1,12 +1,31 @@
...
-- name: Deploy Application
-  hosts: fs
+- name: Deploy An Application Build
+  hosts: fpm:web:fs
   become: true
   any_errors_fatal: true
   pre_tasks:
     - include: tasks/all.yml mode="pre"
+
+    - name: Set build code base fact
```
Comment (clwells) [inline playbooks/deploy.yml]: title case

File: playbooks/tasks/build/sync.yml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+- name: Create Build Code Base
+  file: 
+    path="{{ hostvars[groups['fpm'][0]]['nex_app_build_code_base'] }}"
+    state="directory"
+    mode="0711"
+    owner="{{ nex_app_unix_user }}"
+    group="{{ nex_app_unix_group }}"
+
+- name: Fix build base context
```
Comment (clwells) [inline playbooks/tasks/build/sync.yml]: title case

File: playbooks/tasks/deploy/build.yml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+- name: Application Stop
+  script:
+    application-stop.sh
+
+- name: Unmount
+  script:
+    unmount-app-root.sh {{ nex_app_root }}
+  register: result
+
+- name: Mount build id
```
Comment (clwells) [inline playbooks/tasks/deploy/build.yml]: title case

File: playbooks/tasks/deploy/build.yml [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+#    dest="{{ nfs_shared_path }}/{{ item }}"
+#    recursive="yes"
+#    owner="yes"
+#    perms="yes"
+#  delegate_to: "{{ inventory_hostname }}"
+#  with_items: "{{ nex_app_shared_mounts }}"
+#  when:
+#    - inventory_hostname == groups['fpm'][0]
+
+- name: Sync the shared dirs
```
Comment (clwells) [inline playbooks/tasks/deploy/build.yml]: title case

File: playbooks/tasks/deploy/build.yml [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  script:
+    sync-shared-dir.sh {{ nex_app_root }}/{{ item }} {{ nfs_shared_path }}/{{ item }}
+  delegate_to: "{{ inventory_hostname }}"
+  with_items: "{{ nex_app_shared_mounts }}"
+  when:
+    - inventory_hostname == groups['fpm'][0] 
+
+- name: Mount the shared dirs
```
Comment (clwells) [inline playbooks/tasks/deploy/build.yml]: title case

File: playbooks/_buildapp.yml (hunk 1 above)
Comment (nexpoehler) [inline playbooks/_buildapp.yml]: Good point.  Maybe we'll need to variable-ize the "worker" group name, which this case is fpm.

File: playbooks/files/run-build-hook.sh (hunk 2 above)
Comment (nexpoehler) [inline playbooks/files/run-build-hook.sh]: Yes, although we don't have a "nice" way to communicate hook error messages back to the user yet, which will be important, probably pretty soon.  Getting friendly output from the result of an ansible run (task, or whole playbook) is a thing I haven't figured out a good way to do yet.

File: playbooks/tasks/deploy/build.yml [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+#  synchronize:
+#    src="{{ nex_app_root }}/{{ item }}"
+#    dest="{{ nfs_shared_path }}/{{ item }}"
+#    recursive="yes"
+#    owner="yes"
+#    perms="yes"
+#  delegate_to: "{{ inventory_hostname }}"
+#  with_items: "{{ nex_app_shared_mounts }}"
+#  when:
+#    - inventory_hostname == groups['fpm'][0]
```
Comment (nexpoehler) [inline playbooks/tasks/deploy/build.yml]: It's worth nothing that this synchronize module *should* work but there's a bug currently in ansible sync'ing two directories within a single remote host.  I think it's this issue, which is fixed in ansible dev but not in 2.2.x yet.
https://github.com/ansible/ansible/issues/17097

Comment (nexpoehler) [inline playbooks/tasks/deploy/build.yml]: And, it's commented out to remind me to test with it in the future so the script version below can go away.

**Pull request merged**